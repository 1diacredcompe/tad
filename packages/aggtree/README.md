# `aggtree`

> TODO: description

## Usage

```
const aggtree = require('aggtree');

// TODO: DEMONSTRATE API
```
