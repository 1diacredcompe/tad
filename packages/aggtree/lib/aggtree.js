'use strict';

module.exports = aggtree;

function aggtree() {
    // TODO
}
