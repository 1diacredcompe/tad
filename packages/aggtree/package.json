{
  "name": "aggtree",
  "version": "0.10.0",
  "description": "pivot table built on top of reltab",
  "author": "Antony Courtney <antony.courtney@gmail.com>",
  "homepage": "https://github.com/antonycourtney/tad-pkg#readme",
  "license": "MIT",
  "main": "dist/aggtree.js",
  "types": "dist/aggtree.d.ts",
  "scripts": {
    "prepublish": "npm run build",
    "clean": "rimraf dist",
    "build": "tsc",
    "test": "jest --config jest.config.json"
  },
  "dependencies": {
    "lodash": "^4.17.15",
    "loglevel": "^1.6.8",
    "reltab": "^0.10.0",
    "reltab-sqlite": "^0.10.0"
  },
  "devDependencies": {
    "@types/jest": "^25.2.1",
    "@types/node": "^13.11.0",
    "babel-plugin-flow-to-typescript": "^0.6.0",
    "console.table": "^0.10.0",
    "jest": "^25.2.7",
    "mkdirp": "^1.0.4",
    "prettier": "^2.0.2",
    "ts-jest": "^25.3.1",
    "typescript": "^3.8.3"
  },
  "repository": {
    "type": "git",
    "url": "git+https://github.com/antonycourtney/tad-pkg.git"
  },
  "bugs": {
    "url": "https://github.com/antonycourtney/tad-pkg/issues"
  }
}
