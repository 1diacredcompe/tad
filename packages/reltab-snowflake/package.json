{
  "name": "reltab-snowflake",
  "version": "0.10.0",
  "description": "Snowflake driver for reltab",
  "main": "dist/reltab-snowflake.js",
  "types": "dist/reltab-snowflake.d.ts",
  "scripts": {
    "prepublish": "npm run build",
    "clean": "rimraf dist",
    "build": "tsc",
    "test": "jest --config jest.config.json --no-cache",
    "test-resnap": "jest --config jest.config.json --updateSnapshot"
  },
  "keywords": [
    "relational",
    "sql",
    "database",
    "snowflake"
  ],
  "author": "Antony Courtney <antony@antonycourtney.com>",
  "license": "MIT",
  "dependencies": {
    "loglevel": "^1.7.1",
    "reltab": "^0.10.0",
    "snowflake-sdk": "^1.5.3"
  },
  "devDependencies": {
    "@types/jest": "^25.2.1",
    "@types/node": "^13.11.0",
    "@types/snowflake-sdk": "^1.5.1",
    "aggtree": "^0.10.0",
    "console.table": "^0.10.0",
    "jest": "^25.2.7",
    "mkdirp": "^1.0.4",
    "prettier": "^2.0.2",
    "ts-jest": "^25.3.1",
    "typescript": "^3.8.3"
  }
}
