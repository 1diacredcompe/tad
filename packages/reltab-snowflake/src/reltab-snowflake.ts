import * as log from "loglevel";
import {
  TableRep,
  QueryExp,
  Schema,
  DbConnection,
  defaultEvalQueryOptions,
  EvalQueryOptions,
  DbProvider,
  registerProvider,
} from "reltab";
import {
  TableInfoMap,
  TableInfo,
  Row,
  ColumnMetaMap,
  DbConnectionKey,
  SnowflakeDialect,
  DataSourceNode,
  DataSourcePath,
} from "reltab";

import * as snowflake from "snowflake-sdk";
import * as path from "path";

function safeGetEnv(varName: string): string {
  const val = process.env[varName];
  if (val === undefined) {
    throw new Error(`required environment variable ${varName} is not defined.`);
  }
  return val;
}

export function getAuthConnectionOptions(): snowflake.ConnectionOptions {
  const connOpts: snowflake.ConnectionOptions = {
    account: safeGetEnv("RELTAB_SNOWFLAKE_ACCOUNT"),
    username: safeGetEnv("RELTAB_SNOWFLAKE_USERNAME"),
    password: safeGetEnv("RELTAB_SNOWFLAKE_PASSWORD"),
  };
  return connOpts;
}

function executeQuery(conn: snowflake.Connection, sqlText: string): Promise<any[]> {
  return new Promise((resolve, reject) => {
    let stmt = conn.execute({
      sqlText,
      complete: (err, stmt, rows) => {
        if (err) {
          log.error(
            "Failed to execute statement due to the following error: " +
              err.message
          );
          reject(err);
        } else {
          resolve(rows);
        }
      },
    });
  });
}

function typeBaseName(origType: string): string {
  return origType.split("(", 1)[0];
}

export class SnowflakeConnection implements DbConnection {
  readonly connectionKey: DbConnectionKey;
  tableMap: TableInfoMap;
  snowConn: snowflake.Connection;

  constructor(connectionInfo: snowflake.ConnectionOptions) {
    this.connectionKey = {
      providerName: "snowflake",
      connectionInfo
    };
    this.tableMap = {};
    log.debug(
      "creating snowflake connection with: ",
      JSON.stringify(connectionInfo, null, 2)
    );
    // Enable this for hard-core debugging:
    // snowflake.configure({ logLevel: "TRACE" });
    this.snowConn = snowflake.createConnection(connectionInfo);
  }

  connect(): Promise<void> {
    return new Promise((resolve, reject) => {
      log.debug("connect: about to connect to snowflake...");
      try {
        this.snowConn.connect((err, conn) => {
          if (err) {
            log.error("error connecting to Snowflake: ", err.message);
            log.error(err);
            reject(err);
          } else {
            log.debug("succesfully connected to snowflake");
            resolve();
          }
        });
      } catch (connErr) {
        log.error("caught connect error: ", connErr);
        reject(connErr);
      }
    });
  }

  async getDisplayName(): Promise<string> {
    return "Snowflake";
  }

  async evalQuery(
    query: QueryExp,
    offset?: number,
    limit?: number,
    options?: EvalQueryOptions
  ): Promise<TableRep> {
    let t0 = process.hrtime();
    const schema = query.getSchema(SnowflakeDialect, this.tableMap);
    const sqlQuery = query.toSql(
      SnowflakeDialect,
      this.tableMap,
      offset,
      limit
    );
    let t1 = process.hrtime(t0);
    const [t1s, t1ns] = t1;

    const trueOptions = options ? options : defaultEvalQueryOptions;

    if (trueOptions.showQueries) {
      log.debug("time to generate sql: %ds %dms", t1s, t1ns / 1e6);
      log.debug("SqliteContext.evalQuery: evaluating:");
      log.info(sqlQuery);
    }

    const t2 = process.hrtime();
    const qres = await executeQuery(this.snowConn, sqlQuery);
    log.trace("evalQuery: query results: ", JSON.stringify(qres, null, 2));
    const rows = qres as Row[];
    const t3 = process.hrtime(t2);
    const [t3s, t3ns] = t3;
    const t4pre = process.hrtime();
    const ret = new TableRep(schema, rows);
    const t4 = process.hrtime(t4pre);
    const [t4s, t4ns] = t4;

    if (trueOptions.showQueries) {
      log.info("time to run query: %ds %dms", t3s, t3ns / 1e6);
      log.info("time to mk table rep: %ds %dms", t4s, t4ns / 1e6);
    }

    return ret;
  }

  async rowCount(query: QueryExp, options?: EvalQueryOptions): Promise<number> {
    let t0 = process.hrtime();
    const countSql = query.toCountSql(SnowflakeDialect, this.tableMap);
    let t1 = process.hrtime(t0);
    const [t1s, t1ns] = t1;

    const trueOptions = options ? options : defaultEvalQueryOptions;

    if (trueOptions.showQueries) {
      log.debug("time to generate sql: %ds %dms", t1s, t1ns / 1e6);
      log.debug("SqliteContext.evalQuery: evaluating:");
      log.info(countSql);
    }

    const t2 = process.hrtime();
    const qres = await executeQuery(this.snowConn, countSql);
    const dbRows = qres as Row[];
    const t3 = process.hrtime(t2);
    const [t3s, t3ns] = t3;
    log.debug("time to run query: %ds %dms", t3s, t3ns / 1e6);
    const ret = dbRows[0].rowCount as number;
    return ret;
  }

  async importCsv(): Promise<void> {
    throw new Error("importCsv not yet implemented for Snowflake");
  }

  private async dbGetTableInfo(tableName: string): Promise<TableInfo> {
    const sqlQuery = `DESCRIBE TABLE ${tableName}`;

    const qres = await executeQuery(this.snowConn, sqlQuery);
    const metaRows = qres as Row[];

    const extendCMap = (cmm: ColumnMetaMap, item: any): ColumnMetaMap => {
      const cnm = item.name as string;
      const cType = typeBaseName(item.type as string);

      const cmd = {
        displayName: cnm,
        columnType: cType,
      };
      cmm[cnm] = cmd;
      return cmm;
    };

    const columnIds = metaRows.map((item: any) => item.name);
    const cmMap = metaRows.reduce(extendCMap, {});
    const schema = new Schema(SnowflakeDialect, columnIds, cmMap);
    return {
      tableName,
      schema,
    };
  }

  async getTableInfo(tableName: string): Promise<TableInfo> {
    let ti = this.tableMap[tableName];
    if (!ti) {
      ti = await this.dbGetTableInfo(tableName);
      if (ti) {
        this.tableMap[tableName] = ti;
      }
    }
    return ti;
  }

  async getSourceInfo(path: DataSourcePath): Promise<DataSourceNode> {
    throw new Error("getSourceInfo not yet implemented for Snowflake");
  }
}

const snowflakeDbProvider: DbProvider = {
  providerName: "snowflake",
  connect: async (connectionInfo: any): Promise<DbConnection> => {
    const conn = new SnowflakeConnection(connectionInfo);
    await conn.connect();
    return conn;
  },
};

registerProvider(snowflakeDbProvider);
