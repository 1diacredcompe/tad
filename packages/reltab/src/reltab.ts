export * from "./dialectRegistry";
export * from "./Connection";
export * from "./defs";
export * from "./FilterExp";
export * from "./QueryExp";
export * from "./Schema";
export * from "./TableRep";
export * from "./AggFn";
export * from "./ColumnType";
export * from "./DataSource";
