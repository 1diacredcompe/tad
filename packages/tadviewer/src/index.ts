/*
export * from "./components/AggPanel";
export * from "./components/ColumnList";
export * from "./components/ColumnRow";
export * from "./components/ColumnSelector";
export * from "./components/DisplayOrderPanel";
export * from "./components/FilterEditor";
export * from "./components/FilterEditorRow";
export * from "./components/Footer";
export * from "./components/FormatPanel";
export * from "./components/GridPane";
export * from "./components/IndeterminateCheckbox";
export * from "./components/LoadingModal";
export * from "./components/NumFormatPanel";
export * from "./components/PivotOrderPanel";
export * from "./components/Sidebar";
export * from "./components/SingleColumnSelect";
export * from "./components/SortOrderPanel";
export * from "./components/TextFormatPanel";
*/

export * from "./AppState";
export * from "./components/AppPane";
