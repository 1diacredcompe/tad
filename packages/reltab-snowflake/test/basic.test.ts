import * as reltab from "reltab";
import { SnowflakeConnection, getAuthConnectionOptions } from "../src/reltab-snowflake";
import "../src/reltab-snowflake";
import * as util from "./testUtils";
import * as log from "loglevel";
import * as aggtree from "aggtree";
import { PathTree } from "aggtree";
import { DbConnectionKey } from "reltab";
import * as snowflake from "snowflake-sdk";

let testCtx: SnowflakeConnection;

let connOpts = getAuthConnectionOptions();
connOpts.database = "CITIBIKE";
connOpts.schema = "PUBLIC";

const snowflakeConnKey: DbConnectionKey = {
  providerName: "snowflake",
  connectionInfo: connOpts,
};

beforeAll(async () => {
  // log.setLevel(log.levels.DEBUG);

  testCtx = (await reltab.getConnection(
    snowflakeConnKey
  )) as SnowflakeConnection;
  log.debug("got testCtx: ", testCtx);
  const ti = await testCtx.getTableInfo("TRIPS");
  log.debug("trips table info: ", ti);
});

const tripsTableQuery = reltab.tableQuery("TRIPS");

test("t2 - basic table query", async () => {
  const qres = await testCtx.evalQuery(tripsTableQuery, 0, 20);

  util.logTable(qres, { maxRows: 10 });

  expect(qres).toMatchSnapshot();
});

test("basic rowcount", async () => {
  const rowCount = await testCtx.rowCount(tripsTableQuery);

  console.log("row count: ", rowCount);
});
