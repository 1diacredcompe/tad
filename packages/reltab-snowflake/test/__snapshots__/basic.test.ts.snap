// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`t2 - basic table query 1`] = `
TableRep {
  "rowData": Array [
    Object {
      "BIKEID": 20564,
      "BIRTH_YEAR": 1970,
      "END_STATION_ID": 505,
      "END_STATION_LATITUDE": 40.74901271,
      "END_STATION_LONGITUDE": -73.98848395,
      "END_STATION_NAME": "6 Ave & W 33 St",
      "GENDER": 2,
      "MEMBERSHIP_TYPE": null,
      "STARTTIME": 2013-12-11T09:45:19.000Z,
      "START_STATION_ID": 153,
      "START_STATION_LATITUDE": 40.752062307,
      "START_STATION_LONGITUDE": -73.981632404,
      "START_STATION_NAME": "E 40 St & 5 Ave",
      "STOPTIME": 2013-12-11T09:50:57.000Z,
      "TRIPDURATION": 338,
      "USERTYPE": "Subscriber",
    },
    Object {
      "BIKEID": 19720,
      "BIRTH_YEAR": 1987,
      "END_STATION_ID": 521,
      "END_STATION_LATITUDE": 40.75044999,
      "END_STATION_LONGITUDE": -73.99481051,
      "END_STATION_NAME": "8 Ave & W 31 St",
      "GENDER": 2,
      "MEMBERSHIP_TYPE": null,
      "STARTTIME": 2013-12-11T09:45:22.000Z,
      "START_STATION_ID": 435,
      "START_STATION_LATITUDE": 40.74173969,
      "START_STATION_LONGITUDE": -73.99415556,
      "START_STATION_NAME": "W 21 St & 6 Ave",
      "STOPTIME": 2013-12-11T09:51:01.000Z,
      "TRIPDURATION": 339,
      "USERTYPE": "Subscriber",
    },
    Object {
      "BIKEID": 20879,
      "BIRTH_YEAR": 1985,
      "END_STATION_ID": 284,
      "END_STATION_LATITUDE": 40.739016912,
      "END_STATION_LONGITUDE": -74.00263761,
      "END_STATION_NAME": "Greenwich Ave & 8 Ave",
      "GENDER": 1,
      "MEMBERSHIP_TYPE": null,
      "STARTTIME": 2013-12-11T09:45:25.000Z,
      "START_STATION_ID": 353,
      "START_STATION_LATITUDE": 40.68539567,
      "START_STATION_LONGITUDE": -73.97431458,
      "START_STATION_NAME": "S Portland Ave & Hanson Pl",
      "STOPTIME": 2013-12-11T10:15:25.000Z,
      "TRIPDURATION": 1800,
      "USERTYPE": "Subscriber",
    },
    Object {
      "BIKEID": 15750,
      "BIRTH_YEAR": 1986,
      "END_STATION_ID": 517,
      "END_STATION_LATITUDE": 40.75149263,
      "END_STATION_LONGITUDE": -73.97798848,
      "END_STATION_NAME": "Pershing Square S",
      "GENDER": 1,
      "MEMBERSHIP_TYPE": null,
      "STARTTIME": 2013-12-11T09:45:25.000Z,
      "START_STATION_ID": 379,
      "START_STATION_LATITUDE": 40.749156,
      "START_STATION_LONGITUDE": -73.9916,
      "START_STATION_NAME": "W 31 St & 7 Ave",
      "STOPTIME": 2013-12-11T09:55:09.000Z,
      "TRIPDURATION": 584,
      "USERTYPE": "Subscriber",
    },
    Object {
      "BIKEID": 15009,
      "BIRTH_YEAR": 1987,
      "END_STATION_ID": 161,
      "END_STATION_LATITUDE": 40.72917025,
      "END_STATION_LONGITUDE": -73.99810231,
      "END_STATION_NAME": "LaGuardia Pl & W 3 St",
      "GENDER": 1,
      "MEMBERSHIP_TYPE": null,
      "STARTTIME": 2013-12-11T09:45:27.000Z,
      "START_STATION_ID": 531,
      "START_STATION_LATITUDE": 40.71893904,
      "START_STATION_LONGITUDE": -73.99266288,
      "START_STATION_NAME": "Forsyth St & Broome St",
      "STOPTIME": 2013-12-11T09:51:16.000Z,
      "TRIPDURATION": 349,
      "USERTYPE": "Subscriber",
    },
    Object {
      "BIKEID": 19093,
      "BIRTH_YEAR": 1977,
      "END_STATION_ID": 128,
      "END_STATION_LATITUDE": 40.72710258,
      "END_STATION_LONGITUDE": -74.00297088,
      "END_STATION_NAME": "MacDougal St & Prince St",
      "GENDER": 1,
      "MEMBERSHIP_TYPE": null,
      "STARTTIME": 2013-12-11T09:45:28.000Z,
      "START_STATION_ID": 472,
      "START_STATION_LATITUDE": 40.7457121,
      "START_STATION_LONGITUDE": -73.98194829,
      "START_STATION_NAME": "E 32 St & Park Ave",
      "STOPTIME": 2013-12-11T09:59:39.000Z,
      "TRIPDURATION": 851,
      "USERTYPE": "Subscriber",
    },
    Object {
      "BIKEID": 21009,
      "BIRTH_YEAR": 1983,
      "END_STATION_ID": 280,
      "END_STATION_LATITUDE": 40.73331967,
      "END_STATION_LONGITUDE": -73.99510132,
      "END_STATION_NAME": "E 10 St & 5 Ave",
      "GENDER": 2,
      "MEMBERSHIP_TYPE": null,
      "STARTTIME": 2013-12-11T09:45:37.000Z,
      "START_STATION_ID": 359,
      "START_STATION_LATITUDE": 40.75510267,
      "START_STATION_LONGITUDE": -73.97498696,
      "START_STATION_NAME": "E 47 St & Park Ave",
      "STOPTIME": 2013-12-11T09:49:53.000Z,
      "TRIPDURATION": 256,
      "USERTYPE": "Subscriber",
    },
    Object {
      "BIKEID": 14787,
      "BIRTH_YEAR": 1957,
      "END_STATION_ID": 315,
      "END_STATION_LATITUDE": 40.70355377,
      "END_STATION_LONGITUDE": -74.00670227,
      "END_STATION_NAME": "South St & Gouverneur Ln",
      "GENDER": 2,
      "MEMBERSHIP_TYPE": null,
      "STARTTIME": 2013-12-11T09:45:38.000Z,
      "START_STATION_ID": 307,
      "START_STATION_LATITUDE": 40.71427487,
      "START_STATION_LONGITUDE": -73.98990025,
      "START_STATION_NAME": "Canal St & Rutgers St",
      "STOPTIME": 2013-12-11T09:57:30.000Z,
      "TRIPDURATION": 712,
      "USERTYPE": "Subscriber",
    },
    Object {
      "BIKEID": 19541,
      "BIRTH_YEAR": 1979,
      "END_STATION_ID": 367,
      "END_STATION_LATITUDE": 40.75828065,
      "END_STATION_LONGITUDE": -73.97069431,
      "END_STATION_NAME": "E 53 St & Lexington Ave",
      "GENDER": 1,
      "MEMBERSHIP_TYPE": null,
      "STARTTIME": 2013-12-11T09:45:38.000Z,
      "START_STATION_ID": 457,
      "START_STATION_LATITUDE": 40.76695317,
      "START_STATION_LONGITUDE": -73.98169333,
      "START_STATION_NAME": "Broadway & W 58 St",
      "STOPTIME": 2013-12-11T09:53:31.000Z,
      "TRIPDURATION": 473,
      "USERTYPE": "Subscriber",
    },
    Object {
      "BIKEID": 21087,
      "BIRTH_YEAR": 1963,
      "END_STATION_ID": 434,
      "END_STATION_LATITUDE": 40.74317449,
      "END_STATION_LONGITUDE": -74.00366443,
      "END_STATION_NAME": "9 Ave & W 18 St",
      "GENDER": 1,
      "MEMBERSHIP_TYPE": null,
      "STARTTIME": 2013-12-11T09:45:44.000Z,
      "START_STATION_ID": 153,
      "START_STATION_LATITUDE": 40.752062307,
      "START_STATION_LONGITUDE": -73.981632404,
      "START_STATION_NAME": "E 40 St & 5 Ave",
      "STOPTIME": 2013-12-11T10:05:53.000Z,
      "TRIPDURATION": 1209,
      "USERTYPE": "Subscriber",
    },
    Object {
      "BIKEID": 17306,
      "BIRTH_YEAR": 1970,
      "END_STATION_ID": 404,
      "END_STATION_LATITUDE": 40.7405826,
      "END_STATION_LONGITUDE": -74.00550867,
      "END_STATION_NAME": "9 Ave & W 14 St",
      "GENDER": 1,
      "MEMBERSHIP_TYPE": null,
      "STARTTIME": 2013-12-11T09:45:48.000Z,
      "START_STATION_ID": 475,
      "START_STATION_LATITUDE": 40.73524276,
      "START_STATION_LONGITUDE": -73.98758561,
      "START_STATION_NAME": "E 16 St & Irving Pl",
      "STOPTIME": 2013-12-11T09:55:34.000Z,
      "TRIPDURATION": 586,
      "USERTYPE": "Subscriber",
    },
    Object {
      "BIKEID": 15597,
      "BIRTH_YEAR": 1941,
      "END_STATION_ID": 527,
      "END_STATION_LATITUDE": 40.74315566,
      "END_STATION_LONGITUDE": -73.97434726,
      "END_STATION_NAME": "E 33 St & 1 Ave",
      "GENDER": 1,
      "MEMBERSHIP_TYPE": null,
      "STARTTIME": 2013-12-11T09:45:50.000Z,
      "START_STATION_ID": 368,
      "START_STATION_LATITUDE": 40.73038599,
      "START_STATION_LONGITUDE": -74.00214988,
      "START_STATION_NAME": "Carmine St & 6 Ave",
      "STOPTIME": 2013-12-11T10:08:15.000Z,
      "TRIPDURATION": 1345,
      "USERTYPE": "Subscriber",
    },
    Object {
      "BIKEID": 16987,
      "BIRTH_YEAR": 1977,
      "END_STATION_ID": 510,
      "END_STATION_LATITUDE": 40.7606597,
      "END_STATION_LONGITUDE": -73.98042047,
      "END_STATION_NAME": "W 51 St & 6 Ave",
      "GENDER": 1,
      "MEMBERSHIP_TYPE": null,
      "STARTTIME": 2013-12-11T09:45:55.000Z,
      "START_STATION_ID": 438,
      "START_STATION_LATITUDE": 40.72779126,
      "START_STATION_LONGITUDE": -73.98564945,
      "START_STATION_NAME": "St Marks Pl & 1 Ave",
      "STOPTIME": 2013-12-11T10:04:41.000Z,
      "TRIPDURATION": 1126,
      "USERTYPE": "Subscriber",
    },
    Object {
      "BIKEID": 21109,
      "BIRTH_YEAR": 1976,
      "END_STATION_ID": 465,
      "END_STATION_LATITUDE": 40.75513557,
      "END_STATION_LONGITUDE": -73.98658032,
      "END_STATION_NAME": "Broadway & W 41 St",
      "GENDER": 1,
      "MEMBERSHIP_TYPE": null,
      "STARTTIME": 2013-12-11T09:45:59.000Z,
      "START_STATION_ID": 528,
      "START_STATION_LATITUDE": 40.74290902,
      "START_STATION_LONGITUDE": -73.97706058,
      "START_STATION_NAME": "2 Ave & E 31 St",
      "STOPTIME": 2013-12-11T09:57:07.000Z,
      "TRIPDURATION": 668,
      "USERTYPE": "Subscriber",
    },
    Object {
      "BIKEID": 15761,
      "BIRTH_YEAR": 1991,
      "END_STATION_ID": 388,
      "END_STATION_LATITUDE": 40.749717753,
      "END_STATION_LONGITUDE": -74.002950346,
      "END_STATION_NAME": "W 26 St & 10 Ave",
      "GENDER": 2,
      "MEMBERSHIP_TYPE": null,
      "STARTTIME": 2013-12-11T09:46:05.000Z,
      "START_STATION_ID": 229,
      "START_STATION_LATITUDE": 40.72743423,
      "START_STATION_LONGITUDE": -73.99379025,
      "START_STATION_NAME": "Great Jones St",
      "STOPTIME": 2013-12-11T10:02:52.000Z,
      "TRIPDURATION": 1007,
      "USERTYPE": "Subscriber",
    },
    Object {
      "BIKEID": 15412,
      "BIRTH_YEAR": 1960,
      "END_STATION_ID": 375,
      "END_STATION_LATITUDE": 40.72679454,
      "END_STATION_LONGITUDE": -73.99695094,
      "END_STATION_NAME": "Mercer St & Bleecker St",
      "GENDER": 1,
      "MEMBERSHIP_TYPE": null,
      "STARTTIME": 2013-12-11T09:46:13.000Z,
      "START_STATION_ID": 280,
      "START_STATION_LATITUDE": 40.73331967,
      "START_STATION_LONGITUDE": -73.99510132,
      "START_STATION_NAME": "E 10 St & 5 Ave",
      "STOPTIME": 2013-12-11T09:50:25.000Z,
      "TRIPDURATION": 252,
      "USERTYPE": "Subscriber",
    },
    Object {
      "BIKEID": 16006,
      "BIRTH_YEAR": 1970,
      "END_STATION_ID": 254,
      "END_STATION_LATITUDE": 40.73532427,
      "END_STATION_LONGITUDE": -73.99800419,
      "END_STATION_NAME": "W 11 St & 6 Ave",
      "GENDER": 1,
      "MEMBERSHIP_TYPE": null,
      "STARTTIME": 2013-12-11T09:46:19.000Z,
      "START_STATION_ID": 223,
      "START_STATION_LATITUDE": 40.73781509,
      "START_STATION_LONGITUDE": -73.99994661,
      "START_STATION_NAME": "W 13 St & 7 Ave",
      "STOPTIME": 2013-12-11T09:55:14.000Z,
      "TRIPDURATION": 535,
      "USERTYPE": "Subscriber",
    },
    Object {
      "BIKEID": 17551,
      "BIRTH_YEAR": 1957,
      "END_STATION_ID": 274,
      "END_STATION_LATITUDE": 40.68691865,
      "END_STATION_LONGITUDE": -73.976682,
      "END_STATION_NAME": "Lafayette Ave & Fort Greene Pl",
      "GENDER": 1,
      "MEMBERSHIP_TYPE": null,
      "STARTTIME": 2013-12-11T09:46:21.000Z,
      "START_STATION_ID": 239,
      "START_STATION_LATITUDE": 40.69196566,
      "START_STATION_LONGITUDE": -73.9813018,
      "START_STATION_NAME": "Willoughby St & Fleet St",
      "STOPTIME": 2013-12-11T09:50:11.000Z,
      "TRIPDURATION": 230,
      "USERTYPE": "Subscriber",
    },
    Object {
      "BIKEID": 20242,
      "BIRTH_YEAR": 1960,
      "END_STATION_ID": 388,
      "END_STATION_LATITUDE": 40.749717753,
      "END_STATION_LONGITUDE": -74.002950346,
      "END_STATION_NAME": "W 26 St & 10 Ave",
      "GENDER": 1,
      "MEMBERSHIP_TYPE": null,
      "STARTTIME": 2013-12-11T09:46:23.000Z,
      "START_STATION_ID": 490,
      "START_STATION_LATITUDE": 40.751551,
      "START_STATION_LONGITUDE": -73.993934,
      "START_STATION_NAME": "8 Ave & W 33 St",
      "STOPTIME": 2013-12-11T09:53:44.000Z,
      "TRIPDURATION": 441,
      "USERTYPE": "Subscriber",
    },
    Object {
      "BIKEID": 17961,
      "BIRTH_YEAR": 1973,
      "END_STATION_ID": 546,
      "END_STATION_LATITUDE": 40.74444921,
      "END_STATION_LONGITUDE": -73.98303529,
      "END_STATION_NAME": "E 30 St & Park Ave S",
      "GENDER": 1,
      "MEMBERSHIP_TYPE": null,
      "STARTTIME": 2013-12-11T09:46:27.000Z,
      "START_STATION_ID": 475,
      "START_STATION_LATITUDE": 40.73524276,
      "START_STATION_LONGITUDE": -73.98758561,
      "START_STATION_NAME": "E 16 St & Irving Pl",
      "STOPTIME": 2013-12-11T09:52:02.000Z,
      "TRIPDURATION": 335,
      "USERTYPE": "Subscriber",
    },
  ],
  "schema": Object {
    "columnMetadata": Object {
      "BIKEID": Object {
        "columnType": "NUMBER",
        "displayName": "BIKEID",
      },
      "BIRTH_YEAR": Object {
        "columnType": "NUMBER",
        "displayName": "BIRTH_YEAR",
      },
      "END_STATION_ID": Object {
        "columnType": "NUMBER",
        "displayName": "END_STATION_ID",
      },
      "END_STATION_LATITUDE": Object {
        "columnType": "FLOAT",
        "displayName": "END_STATION_LATITUDE",
      },
      "END_STATION_LONGITUDE": Object {
        "columnType": "FLOAT",
        "displayName": "END_STATION_LONGITUDE",
      },
      "END_STATION_NAME": Object {
        "columnType": "VARCHAR",
        "displayName": "END_STATION_NAME",
      },
      "GENDER": Object {
        "columnType": "NUMBER",
        "displayName": "GENDER",
      },
      "MEMBERSHIP_TYPE": Object {
        "columnType": "VARCHAR",
        "displayName": "MEMBERSHIP_TYPE",
      },
      "STARTTIME": Object {
        "columnType": "TIMESTAMP_NTZ",
        "displayName": "STARTTIME",
      },
      "START_STATION_ID": Object {
        "columnType": "NUMBER",
        "displayName": "START_STATION_ID",
      },
      "START_STATION_LATITUDE": Object {
        "columnType": "FLOAT",
        "displayName": "START_STATION_LATITUDE",
      },
      "START_STATION_LONGITUDE": Object {
        "columnType": "FLOAT",
        "displayName": "START_STATION_LONGITUDE",
      },
      "START_STATION_NAME": Object {
        "columnType": "VARCHAR",
        "displayName": "START_STATION_NAME",
      },
      "STOPTIME": Object {
        "columnType": "TIMESTAMP_NTZ",
        "displayName": "STOPTIME",
      },
      "TRIPDURATION": Object {
        "columnType": "NUMBER",
        "displayName": "TRIPDURATION",
      },
      "USERTYPE": Object {
        "columnType": "VARCHAR",
        "displayName": "USERTYPE",
      },
    },
    "columns": Array [
      "TRIPDURATION",
      "STARTTIME",
      "STOPTIME",
      "START_STATION_ID",
      "START_STATION_NAME",
      "START_STATION_LATITUDE",
      "START_STATION_LONGITUDE",
      "END_STATION_ID",
      "END_STATION_NAME",
      "END_STATION_LATITUDE",
      "END_STATION_LONGITUDE",
      "BIKEID",
      "MEMBERSHIP_TYPE",
      "USERTYPE",
      "BIRTH_YEAR",
      "GENDER",
    ],
    "dialect": "snowflake",
  },
}
`;
